blueprints/refetch-invalidation.md

```diff
@@ -0,0 +1,282 @@
+		"id":1,
+		"keyword":"PURGE",
+		"parameters":"TTL:24h""startTime":"2021-06-02 09:23:21-06"
+	}
+}
+```
+
+Sample new response (includes the `invalidationType` on parameters field):
+```json
+{
+	"alerts":[
+		{
+			"text":"Invalidation request created for http://amc-linear-origin.local.tld/path/.*\\.jpeg, start:2021-06-02 15:23:21.348 +0000 UTC end 2021-06-03 15:23:21.348 +0000 UTC",
+			"level":"success"
+		}
+	],
+	"response":{
+		"assetUrl":"http://amc-linear-origin.local.tld/path/.*\\.jpeg",
+		"createdBy":"admin",
+		"deliveryService":"amc-live",
+		"id":1,
+		"keyword":"PURGE",
+		"parameters":"TTL:24h,invalidationType:refresh",
+		"startTime":"2021-06-02 09:23:21-06"
+	}
+}
+```
+
+___
+
+> Note: There are still 1.x routes that reference `UserInvalidationJob`, such as 
+		`user/current/jobs(/|\.json/?)?$`
+		`user/current/jobs(/|\.json/?)?$`
+		These routes are currently deprecated and the corresponding `structs` will be removed in a future release as well.
+
+#### Client Impact
+
+Likewise with Traffic Portal, the `go` clients will need to be updated to provide this additional functionality. Since an additional field has been added to `InvalidationJobInput` in `go-tc` lib, this can be set by the client as well. If left unset, it will default to "refresh".
+
+#### Data Model / Database Impact
+
+
+The current column `parameters` will now contain a cskv (comma separated key value) string. Currently it only stores the `TTL` for the invalidation request:
+```
+TTL:48h
+```
+
+Moving forward, this column will also contain the type of cache invalidation. For instance, the string may read:
+```
+TTL:48h,invalidationType:refetch
+```
+
+If there is no `invalidationType` in the cskv, it is assumed to be **REFRESH**/**STALE** as it's default value keeping with the current implementation. Otherwise the `invalidationType` will be either `refetch` or `refresh`, defaulting to `refresh` during validation.
+
+*OPTIONAL, BUT RECOMMENDED*: As part of this effort, the _Boy Scout Rule_ will be applied ("Always leave the campground cleaner than you found it."). The `agent`, `status`, `asset_type`, `object_type`, and `object_name` columns will be removed. `agent` and `status` are currently hardcoded to the value of 1. Similarly, `asset_type` is never a anything besides "file". `object_type` and `object_name` are not used at all. This would require a DB migration with Goose as well. If this were not implemented, no migration is necessary.
+
+> The removal of these columns will impact the `UserInvalidationJob` struct. Even though the only endpoint utilizing this struct (v1.1) is deprecated, it is still in use.
+
+### ORT/T3C Impact
+
+The `regexrevalidatedotconfig.go` in _lib/tc-atscfg_ currently has a function called `filterJobs` that is responsible for parsing the _parameters_ information from the job. Right now, that is only parsing the **TTL**, however it would then also need to parse the optional, extra field of **invalidationType**. 
```

Comment:
What considerations are made for TrafficServer version dependencies? 

Is `MISS` supported by all ATS versions? If not, what happens when an operator sets a `REFETCH` job on a set of caches that do not support that invalidation type?

Reply:
Seems like there are two options when `MISS` is unavailable based on ATS version: ignore it or use refresh. Can you suggest which would be better?

Reply:
I'm not an ATS expert, so I can't make a solid suggestion here. 

I think whatever happens probably needs to be bubbled up to the operator. You don't want to silently ignore or silently do a stale-refresh when you expect a Miss-Refetch.

Reply:
Well, when I need an ATS and/or ORT/t3c/atstccfg expert I ask @rob05c - lets see if he has an opinion

Reply:
It's definitely not supported on older versions. It's a new feature of the regex_revalidate plugin.

1. Unfortunately, there's not an easy way to bubble it up. We don't really have a mechanism to send data from caches to TO, and even if we did, it'd be pretty unwieldy to collect that from every single cache.

We do have the ATS version in a Parameter in TO, we could hard-code it. But someone might backport the plugin changeset to an older version, and then ATC just won't allow using it. So that's not a great solution.

2. the "refetch" feature is **only** necessary or useful if the Origin server is fundamentally broken and completely violating the HTTP Spec. Without "refetch," ATS does an IMS request. So, the only reason it's ever needed is when a Tenant's Origin is responding with a 304 when the asset actually is modified. In complete violation of RFC 7234 and every HTTP Caching specification. But of course, we deal with horribly broken Origins every day. But I do think we should make that clear in the docs, and maybe the UI.

3. To work correctly, every single Cache has to support it. Which is mostly fine, since CDN operators don't _usually_ operate with widely varying ATS versions. Some do, but I think the majority have a single version and maybe some newer Canaries.

With those 3 things in mind, I think the best solution might be to make it a config Feature Flag in Traffic Ops/+Portal, with Documentation that they must have version X of regex_revalidate. As an additional safety, we should also make `t3c` check and log an error (which operators can then Alarm on).

It's not as good as bubbling it up, but considering all those things, that's the best solution I'm seeing at the moment.

Reply:
Just for awareness @TaylorCFrey, a "feature flag" in TO usually means a parameter in the `GLOBAL` profile to enable the new feature -- similar to what we've done in the past w/ `use_tenancy` and what we currently do w/ `use_reval_pending`. That sounds like a good solution here.

Reply:
Could also be in the TO configuration file

Reply:
No, because really the feature flag is for `t3c` since it's generating the ATS config. I guess it could be a `t3c` option, but I'm not sure that's better than our tried and true method of global profile params.

Reply:
Maybe I'm misunderstanding how this is supposed to work. I thought it'd be like: you make a new job, if it's using "refetch" then TO checks if the feature flag is allowed. If it isn't, you can't make that job, and any existing jobs are filtered out of GET requests for the /jobs endpoint.

It sounds like you're saying you set the flag in TO, then ORT/t3c/atstccfg looks at it and determines based on that what to do with jobs that use "refetch" - which makes sense, because otherwise I suppose I don't know how `t3c` could "check and log an error"

Reply:
> It sounds like you're saying you set the flag in TO, then ORT/t3c/atstccfg looks at it and determines based on that what to do with jobs that use "refetch" - which makes sense, because otherwise I suppose I don't know how t3c could "check and log an error"

Yes, that is how `use_reval_pending` works currently and is how I imagine a feature flag for this feature would work as well.

Reply:
When I said "feature flag" I was thinking the TO configuration file; but a global Parameter also works.

> I thought it'd be like: you make a new job, if it's using "refetch" then TO checks if the feature flag is allowed. If it isn't, you can't make that job

That's what I was thinking, but also having a check in `t3c-apply` as a fail-safe, if it ever does get a job and the installed regex_revalidate is too old to understand it, it would log an error (but probably still do the reval, because an IMS reval is better than nothing and doesn't hurt if they wanted a no-IMS reval).

Reply:
We need to not allow creating those jobs at all, if the feature-flag isn't set. Because by the time it gets to the cache, it's too late: the user put in jobs, that they expected to work, and now we have an error we can't reasonably propagate up to them that the job won't work. 

But we already knew it wasn't going to work, we have the flag in TO. We should be able to return an error from the /jobs POST.

Reply:
Bearing in mind, these won't be intermittent errors. The endpoint will always return an error if the flag isn't set.

That's a good reason to put the flag in the Parameter instead of the config file: the Param is already available in the API. Which lets UIs like the Portal fetch it, to decide whether to hide that extra button from users, so they never even see the option that will always error.

Reply:
I've added some language specifically stating the addition of a global parameter, defaulting to rejecting `refetch` POSTS until specfically set to allow.